### DISCUSSION TITLE
How to raise custom Exception from null validation?

### DISCUSSION THREAD

**Discussion by hiel** [fastapi/fastapi#9638] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
class FooRequest1(BaseModel):
    name: Body(...)

    @root_validator(pre=True)
    def execute(cls, values):
        if values.get("name") is None:
            raise FooException("CUSTOM MESSAGE")
        return values

class FooRequest2(BaseModel):
    name: Query(...)

    @root_validator(pre=True)
    def execute(cls, values):
        if values.get("name") is None:
            raise FooException("CUSTOM MESSAGE")
        return values
```

### Description

I wanna raise custom Exception from request validation.

`FooRequest1` is raises `FooException` fine. when name field is missing. (FooRequest1 works with `Body`)
the name is not null model field (`Body(...)`), but `root_validator` execute first. so it works fine.

but `FooRequest2` is not work. (FooRequest2 works with `Query`)
it raises Pydantic Exception (`RequestValidationError`)
the `root_validator` doesn't execute, because the name field invalid with BaseModel.

why? what's the difference?
is there any other best practice handling validation with custom exception?

### Operating System

macOS

### FastAPI Version

0.87.0

### Python Version

3.10.9

**Comment by esrefzeki** [comment, marked as answer]
The reason why the root_validator is not being executed for FooRequest2 (which uses Query) is because the validation of query parameters in FastAPI is handled by FastAPI's internal mechanisms, not Pydantic's validators.

In FastAPI, when you define a request parameter using Query, it creates a dependency on that parameter and handles the validation internally. If the validation fails, FastAPI automatically raises a FastAPI.HTTPException with the appropriate status code and error message.

If you want to raise a custom exception when the query parameter is missing, you can achieve it by using a different approach. You can define a dependency for your endpoint and perform the validation there. Here's an example:

```
from fastapi import Depends, FastAPI, HTTPException, Query

app = FastAPI()

def validate_name(name: str = Query(...)) -> str:
    if name is None:
        raise FooException("CUSTOM MESSAGE")
    return name

@app.get("/endpoint")
async def endpoint(name: str = Depends(validate_name)):
    # Your endpoint logic here
    return {"name": name}
```

In this example, the validate_name function acts as a dependency for the endpoint route. It receives the query parameter name and performs the validation. If the name is missing or invalid, it raises your custom FooException. Otherwise, the validated name value is passed to the endpoint function.

By using this approach, you can raise your custom exception and handle the validation in a centralized manner for all endpoints that depend on the name query parameter.

**Reply by hiel** [reply]
@esrefzeki 
thanks for comment :)
I have one more question.
then, is there no way to validate like pre root_validator?
1. the field is not null, so swagger have to display it `required`, so the field type have to like this (`Query(...,`), not (`Query(None,`)
2. I wanna raise custom exception when the requested field is null, so the validation function have to execute before initialize Request class (before responsing FastAPI exception message `1 validation error for Request\nquery -> ...`)
3. And I need to validate function to could access all field at once. Query `Depends` solution can access only one field at once.
    I need like this.
   ```
    def valid(search_type, search_value):
        if search_type == 'MEMBER_NO' and type(search_value) is not int:
            raise CusomException
   ```

is there any other way? now I define Query parameter `Query(None,`, and validate None with `root_validator(pre=True)`.
so the Swagger UI display it `not required field` :(

**Comment by esrefzeki** [comment]
Your welcome,

1. Your answer is Pydantic :) When you declare your class fields like:

```
class FooRequest1(BaseModel):
    name: str
    surname: Optional[str]
```

then it will show it on swagger required for name. And optional for surname.

2. You can create if condition to check if any field you want to be not null and warn user any way you want.

3. Use pydantic :) That is the pill.

**Reply by hiel** [reply]
@esrefzeki 
Sorry for the many questions 🙏 

you mean, just use Pydantic like this?
```
class FooRequest1(BaseModel):
    name: str -> swagger required
    sur_name: Optional[str] -> swagger not required
    
    @root_validator(pre=True)
    def validation(cls, values):
        if values.get('name') is None:
            raise CustomException
```

but I have to use `Query`, `Body`, because of `alias` and `description`
```
class FooRequest1(BaseModel):
    name: str = Query(None, alias='some_name', description='description')
    sur_name: Optional[str] = Query(None, alias='surName', description='description')

    @root_validator(pre=True)
    def validation(cls, values):
        if values.get('name') is None:
            raise CustomException
```

case 1
```
name: str = Query(None, alias='some_name', description='description')
raise CustomException fine, but swagger display not required
```

case 2
```
name: str = Query(None, alias='some_name', description='description')
swagger display required fine, but it raise Pydantic Exception, not CustomException.
```

case 3
```
name: str | None = Query(..., alias='some_name', description='description')
swagger display required fine, but it raise Pydantic Exception, not CustomException.
```

case 4
```
name: str | None = Query(None, alias='some_name', description='description')
raise CustomException fine, but swagger display not required
```

I know the `Query`, `Body` is fastapi's class. so I think this discussion is related with fastapi.
if not, I will discuss Pydantic github section :)
